### ISSUE TITLE
Training becomes non-reproducible after loading pretrained weights

### ISSUE THREAD

**Issue by parsec008** [ultralytics/ultralytics#20916]
I fixed the seed and loaded yolov8n.pt, but when I repeat the training, the accuracy varies each time—sometimes by 1–2%. However, if I don’t use pretrained weights, the results are identical every run. Why does this happen, and is there any way to achieve reproducibility when loading pretrained weights?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @parsec008, thank you for bringing this up and for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, where you might find answers to common questions on reproducibility and pretrained weights.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the issue more effectively.

If this is a custom training ❓ Question, please share more details such as your training script, dataset configuration, and logs, and ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community to connect, get help, or share your experience:
- For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- For community stories and Q&A, check out our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

To make sure your issue isn’t already resolved, upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs smoothly in these verified, up-to-date environments (all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, covering all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help you get started—an Ultralytics engineer will review your issue and assist you soon!

**Comment by Y-T-G** [comment]
What's the code you're using?

Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by parsec008** [comment]
My dataset has only three classes, I changed nc to 3 when using yolov8.yaml, which causes the head’s weights to fail to load. When nc is 80, loading the pretrained weights produces the same training results every time. I’m not sure if it’s because the head is randomly initialized, since I’ve already fixed the seed.

Here is my training code:
from ultralytics import YOLO
model = YOLO("ultralytics/cfg/v8/yolov8.yaml").load("yolov8n.pt")
train_results = model.train(
    data="ultralytics/cfg/datasets/new.yaml", 
    epochs=200,  
    imgsz=640,  
    batch=32,
    device="0",
    seed=0
) 
yolo checks:
Ultralytics 8.3.146 🚀 Python-3.12.3 torch-2.3.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24154MiB)
Setup complete ✅ (128 CPUs, 503.1 GB RAM, 10.9/30.0 GB disk)

OS                  Linux-5.15.0-88-generic-x86_64-with-glibc2.35
Environment         Docker
Python              3.12.3
Install             pip
Path                /root/miniconda3/lib/python3.12/site-packages/ultralytics
RAM                 503.14 GB
Disk                10.9/30.0 GB
CPU                 Intel Xeon Platinum 8362 2.80GHz
CPU count           128
GPU                 NVIDIA GeForce RTX 3090, 24154MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.3.0+cu121>=1.8.0
torch               ✅ 2.3.0+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.18.0+cu121>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0

**Comment by Y-T-G** [comment]
You shouldn't modify the `nc` in the YAML. It's updated automatically

**Comment by parsec008** [comment]
I tested it: when nc was set to 80, the trained model's Detect output still had 3 classes. But I want to know why changing nc makes the results non-reproducible. It should only randomly reinitialize the Detect layer, and that part should still be controlled by the seed.

**Comment by glenn-jocher** [comment]
The non-reproducibility occurs because when you modify `nc` in the YAML, the detection head weights can't be loaded from the pretrained model and get randomly reinitialized. Even with a fixed seed, PyTorch's random number generator state can be affected by the order of operations during partial weight loading, causing slight variations in the reinitialized layers.

For reproducible training with custom classes, load the pretrained model directly without modifying the YAML:

```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")  # Load pretrained directly
results = model.train(data="path/to/your/data.yaml", epochs=200, seed=0)
```

The model will automatically adapt to your dataset's number of classes while maintaining reproducibility since the backbone weights load consistently.
